nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/schema/inference/FieldTypeInference.java

```diff
@@ -62,36 +62,44 @@ public void addPossibleDataType(final DataType dataType) {
             final RecordSchema newSchema = ((RecordDataType) dataType).getChildSchema();
 
             final RecordSchema mergedSchema = DataTypeUtils.merge(singleDataTypeSchema, newSchema);
+            possibleDataTypes.remove(singleDataType);
             singleDataType = RecordFieldType.RECORD.getRecordDataType(mergedSchema);
+            possibleDataTypes.add(singleDataType);
             return;
         }
 
-        if (singleFieldType.isWiderThan(additionalFieldType)) {
-            // Assigned type is already wide enough to encompass the given type
-            return;
+        if (possibleDataTypes.isEmpty()) {
+            possibleDataTypes.add(singleDataType);
         }
 
-        if (additionalFieldType.isWiderThan(singleFieldType)) {
-            // The given type is wide enough to encompass the assigned type. So changed the assigned type to the given type.
-            singleDataType = dataType;
-            return;
-        }
+        boolean hasWiderNonString = possibleDataTypes.stream()
```

Comment:
This method is also evaluated against potentially every field in any record whose schema is being inferred, so performance is important. As such, we should avoid use of `Stream`s and instead prefer imperative style programming.

Reply:
Replaced Stream api usage.